reps/2023-03-15-train-api.md

```diff
@@ -0,0 +1,162 @@
+
+2. Introduce `Dataset.get_logical_plan()` (DeveloperAPI), which returns the logical plan that can be used to extract the lineage of preprocessors applied to this Dataset. If multiple preprocessors are applied, Train can return a `Chain` of the preprocessors. Non-preprocessor operations on the Dataset are ignored, and we can also allow ignoring preprocessors such as per epoch preprocessors. This function will be used by Train to persist fitted preprocessors with checkpoints.
+
+3. Deprecate the following additional Trainer configs when streaming is enabled: `global_shuffle` and `randomize_block_order` (user to use native Dataset shuffle ops), and the preprocessing args `fit`, `transform`, `preprocessor`, and `per_epoch_preprocessor` (user to setup preprocessing explicitly prior to creating Trainer).
+
+## Example:
+
+### Before
+```python
+
+base = ray.data.read_parquet("s3://bucket/etl_output")
+fact_table = ray.data.read_csv("s3://bucket/my.csv")
+
+# Create the preprocessor.
+prep = StandardScaler(["f1", "f2"])
+
+# Create the per-epoch preprocessor.
+per_epoch_prep = RandomNoisePreprocessor()
+
+# Trainer applies preprocessing internally via config.
+trainer = TorchTrainer(
+    model,
+    datasets={
+        "train_ds": train_ds,
+        "fact_table": fact_table,
+    },
+    scaling_config=ScalingConfig(num_workers=4),
+    preprocessor=prep,
+    dataset_config={
+        "train_ds": {
+            "max_object_store_memory_fraction": 0.2,  # Enable streaming.
+            "randomize_block_order": True,
+            "per_epoch_preprocessor": per_epoch_prep,
+        },
+    },
+)
+
+# Checkpoint includes fitted preprocessor.
+best_checkpoint = trainer.fit().checkpoint
+assert best_checkpoint.get_preprocessor() == prep
+```
+
+### After
+```python
+
+base = ray.data.read_parquet("s3://bucket/etl_output")
+fact_table = ray.data.read_csv("s3://bucket/my.csv")
+
+# Fit the preprocessor.
+prep = StandardScaler(["f1", "f2"])
+prep.fit(base)
+
+# Apply base preprocessing.
+train_ds = base.map_batches(prep)
+train_ds.cache()  # Optional: cache the base data in memory.
+
+# Per-epoch preprocessing.
+train_ds = train_ds \
+    .randomize_block_order() \
+    .map_batches(RandomNoisePreprocessor())
+
+# Trainer doesn't know about preprocessing at all.
+trainer = TorchTrainer(
+    model,
+    datasets={
+        "train_ds": train_ds,
+        "fact_table": fact_table,
+    },
+    scaling_config=ScalingConfig(num_workers=4),
+    dataset_config={
+        "train_ds": {
+            "resource_limits": ExecutionResources(
+                object_store_memory=20e9,  # Customized streaming memory limit.
+            ),
+        },
+    },
+)
+
+# Checkpoint includes fitted preprocessor.
+best_checkpoint = trainer.fit().checkpoint
+assert best_checkpoint.get_preprocessor() == prep
+```
+
```

Comment:
can you add an example for training -> inference?

In particular, how would the user specify certain preprocessors to not use for inference?

Reply:
Added as `ignore_for_inference` attribue.